doc/design/logging.md

```diff
@@ -0,0 +1,385 @@
+# Logging
+
+## Summary and problem description
+
+Logging is a crucial part of a framework since it informs the developer and
+user of the framework about anomalies and helps with debugging by providing
+contextual information.
+
+Additionally, when integrated into a separate framework, the log messages should
+be forwarded in order to have a single logging infrastructure.
+
+The logging should also be performant and not allocate memory. Ideally, it should
+be possible to disable it at compile time and let the compiler optimize it away.
+
+It should also be possible to alter the behavior of the logging via environment
+variables, e.g. the log level.
+
+## Terminology
+
+### Log levels and how to use them
+
+| LogLevel | Usage |
+|----------|-------|
+| FATAL    | For fatal and non-recoverable errors which essentially lead to termination of the program. |
+| ERROR    | For severe but recoverable errors. |
+| WARN     | For cases when something unintentional happens which is not considered a severe error. |
+| INFO     | For important status information a user should be aware of. |
+| DEBUG    | Anything that is helpful for debugging. |
+| TRACE    | Any comment could be replaced by logging call with log level trace. |
+
+The log levels `FATAL`, `ERROR` and `WARN` should not be used independently but
```

Comment:
Just because we already had a lengthy discussion about this sentence. Could you add a sentence which also states that the error handler should also be used only in combination with an appropriate log message.

Reply:
I'm not sure if this belongs to this document. Usually I would look at the design document for the error handler for this but I can add a `and vice versa` if that makes you happy :)